compiler/rustc_arena/src/lib.rs

```diff
@@ -270,8 +198,35 @@ impl<T> TypedArena<T> {
 
     #[inline]
     pub fn alloc_from_iter<I: IntoIterator<Item = T>>(&self, iter: I) -> &mut [T] {
+        // This implementation is entirely separate to
+        // `DroplessIterator::alloc_from_iter`, even though conceptually they
+        // are the same.
+        //
+        // `DroplessIterator` (in the fast case) writes elements from the
+        // iterator one at a time into the allocated memory. That's easy
+        // because the elements don't implement `Drop`. But for `TypedArena`
+        // they do implement `Drop`, which means that if the iterator panics we
+        // could end up with some allocated-but-uninitialized elements, which
+        // will then cause UB in `TypedArena::drop`.
+        //
+        // Instead we use an approach where any iterator panic will occur
+        // before the memory is allocated. This function is much less hot than
+        // `DroplessArena::alloc_from_iter`, so it doesn't need to be
+        // hyper-optimized.
```

Comment:
Do we need a fast path if `!needs_drop::<T>()`? That would call the drop-less codepath and eventually benefit from the opt.

Reply:
Not sure I understand the question. Any `T` not implementing `Drop` will be allocated in the `DroplessArena`, as per [this comment](https://github.com/rust-lang/rust/blob/eea26141ec2bf07a825011af841463dec51a04c9/compiler/rustc_arena/src/lib.rs#L603-L615).